src/cpp/src/visual_language/minicpm/classes.cpp

```diff
@@ -575,35 +622,38 @@ ov::Tensor InputsEmbedderMiniCPM::get_inputs_embeds(const std::string& prompt, c
     std::vector<ov::Tensor> single_images = to_single_image_tensors(images);
 
     for (const ov::Tensor& image : single_images) {
-        EncodedImage encoded_image = m_vision_encoder->encode(image);
+        embeds.push_back(m_vision_encoder->encode(image));
+    }
+
+    size_t n_prev_images = m_image_id;
+    std::string native_tag = "<image>./</image>";
+    auto [unified_prompt, images_sequence] = unify_prompt(prompt, native_tag, embeds.size(), n_prev_images);
+
+    for (const EncodedImage& encoded_image : embeds) {
+        std::string expanded_tag;
         if (m_vlm_config.use_image_id) {
-            images_prompt += m_vlm_config.im_id_start + std::to_string(m_image_id) + m_vlm_config.im_id_end;
+            expanded_tag += m_vlm_config.im_id_start + std::to_string(m_image_id) + m_vlm_config.im_id_end;
             ++m_image_id;
         }
         std::string unk64;
         for (size_t idx = 0; idx < m_vlm_config.query_num; ++idx) {
             unk64 += m_vlm_config.unk;
         }
-        images_prompt += m_vlm_config.im_start + unk64 + m_vlm_config.im_end;
+        expanded_tag += m_vlm_config.im_start + unk64 + m_vlm_config.im_end;
         if (encoded_image.slices) {
             ov::Shape slices_shape = encoded_image.slices.get_shape();
             for (size_t row_idx = 0; row_idx < slices_shape.at(0); ++row_idx) {
                 for (size_t col_idx = 0; col_idx < slices_shape.at(1); ++col_idx) {
-                    images_prompt += m_vlm_config.slice_start + unk64 + m_vlm_config.slice_end;
+                    expanded_tag += m_vlm_config.slice_start + unk64 + m_vlm_config.slice_end;
                 }
-                images_prompt += '\n';
+                expanded_tag += '\n';
             }
+            expanded_tag.pop_back();  // Equivalent of python "\n".join(slices).
         }
-        if ('\n' != *(images_prompt.end() - 1)) {
-            // Image wasn't sliced, add \n to the end of image anyway.
-            // Strangely, \n isn't placed between </image><slice>.
-            images_prompt += '\n';
-        }
-        embeds.push_back(std::move(encoded_image));
+        unified_prompt.replace(unified_prompt.find(native_tag), native_tag.length(), expanded_tag);
     }
```

Comment:
potentially, we could also add a function to replace unified tags by expanded tags:
```c++
void resolve_unified_tags(std::string& unified_prompt, const std::regex& native_tag_pattern, std::function<std::string(size_t)> expanded_tag);
```
- `unified_prompt` - unified prompt
- `native_tag_pattern` regex with native tag
- `expanded_tag` function / lambda, which creates expanded tag based on image ID. Image ID is parsed from `unified_prompt`

We can even combine `unify_prompt` and `resolve_unified_tags` to a single function

Reply:
That wouldn't work for phi3_v. It has a native tag, but instead of expansion, the prompt is split by this tag. Image token ids are artificially inserted into this gap using negative values after tokenization.

I don't think there's a reason to unify tag expansion step. It's going to be unique for every model anyway. Moreover tag expansion always comes after `unify_prompt()` so two sequential calls `unify_prompt(); resolve_unified_tags();` are more readable than passing a lambda.